rav1e_js/src/lib.rs

```diff
@@ -7,88 +7,162 @@
+  }
 
-  let cfg = Config::new().with_encoder_config(enc);
-  let mut ctx: Context<u16> = cfg.new_context().unwrap();
-  log!("Instantiated new encoder (config: {:?})", cfg);
+  pub fn debug(&self) -> String {
+    format!("{:?}", self.ctx)
+  }
 
-  let f = ctx.new_frame();
-  log!("Generated new frame: {:?}", f);
+  /// Allocates and returns a new frame.
+  #[allow(non_snake_case)]
+  pub fn newFrame(&self) -> Frame {
+    Frame { f: self.ctx.new_frame() }
+  }
 
-  for i in 0..repeats {
-    log!("Sending frame {}", i);
-    match ctx.send_frame(f.clone()) {
-      Ok(_) => {}
+  /// Sends the frame for encoding.
+  ///
+  /// This method adds the frame into the frame queue and runs the first passes of the look-ahead computation.
+  #[allow(non_snake_case)]
+  pub fn sendFrame(&mut self, frame: &Frame) -> Result<(), JsValue> {
+    match self.ctx.send_frame(frame.f.clone()) {
+      Ok(_) => Ok(()),
       Err(e) => match e {
-        EncoderStatus::EnoughData => {
-          log!("Warn: Unable to append frame {} to the internal queue", i);
-        }
-        _ => {
-          panic!("Unable to send frame {}", i);
-        }
+        EncoderStatus::EnoughData => Err(construct_js_err(
+          e,
+          "Unable to append frame to the internal queue",
+        )),
+        _ => Err(construct_js_err(e, "")),
       },
     }
   }
-  log!("Successfully send {} frames to the encoder", repeats);
-
-  ctx.flush();
-  log!("Flush encoder");
-
-  loop {
-    match ctx.receive_packet() {
-      Ok(packet) => {
-        // Mux the packet.
-        log!("Received packet: {}", packet)
-      }
-      Err(EncoderStatus::Encoded) => {
-        // A frame was encoded without emitting a packet. This is normal,
-        // just proceed as usual.
-      }
-      Err(EncoderStatus::LimitReached) => {
-        // All frames have been encoded. Time to break out of the loop.
-        break;
-      }
-      Err(EncoderStatus::NeedMoreData) => {
-        // The encoder has requested additional frames. Push the next frame
-        // in, or flush the encoder if there are no frames left (on None).
-        ctx.flush();
-      }
-      Err(e) => {
-        panic!(e);
-      }
+
+  /// Flushes the encoder.
+  ///
+  /// Flushing signals the end of the video. After the encoder has been flushed, no additional frames are accepted.
+  pub fn flush(&mut self) {
+    self.ctx.flush();
+  }
+
+  /// Encodes the next frame and returns the encoded data.
+  ///
+  /// This method is where the main encoding work is done.
+  #[allow(non_snake_case)]
+  pub fn encodeNext(&mut self) -> Result<Packet, JsValue> {
+    match self.ctx.receive_packet() {
+      Ok(packet) => Ok(Packet { p: packet }),
+      Err(e) => Err(construct_js_err(e, "")),
     }
   }
-  format!("Done encoding the same tiny blank frame {} times.", repeats)
+
+  // CONFIG OPTIONS
+  //
+
+  /// Set width and height of the frames in pixels.
+  pub fn set_dim(&mut self, width: usize, height: usize) -> Self {
```

Comment:
As mentioned on IRC: You really should have some separation of concerns and not copy around everything every time some setting is updated.